staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go

```diff
@@ -454,6 +456,12 @@ func (c *AvailableConditionController) deleteAPIService(obj interface{}) {
 
 // there aren't very many apiservices, just check them all.
 func (c *AvailableConditionController) getAPIServicesFor(obj runtime.Object) []*apiregistration.APIService {
+	var svc *v1.Service
+	if svc, ok := obj.(*v1.Service); ok {
+		if services, ok := c.cache[svc]; ok {
```

Comment:
using a pointer as a cache key will not work properly

Reply:
@liggitt 
Can we use Service.ObjectMeta.GenerateName as the hash key ?

Reply:
no, generate name is not unique. between services and endpoints, endpoints are far more likely to be updated frequently and cause contention.

I think it would make more sense to maintain a `map[service-namespace]map[service-name][]APIServiceName` based on add/update/delete events from APIServices, since those are very slow-moving. Updating or removing an apiservice object would also need to update that map

That would make the handlers for service and endpoint events an O(1) lookup

Reply:
The lookup is supposed to return []\*apiregistration.APIService - not just APIServiceName.

How about declaring the map as:
map[service-namespace]map[service-name][]\*apiregistration.APIService

Reply:
> The lookup is supposed to return []*apiregistration.APIService - not just APIServiceName.

the only thing done with the results is enqueing, which can be done with the name instead.

> How about declaring the map as:
> map[service-namespace]map[service-name][]*apiregistration.APIService

we don't want to have to update the map if an APIService change doesn't change the referenced service namespace/name. just storing the apiservice name is sufficient

Reply:
I will change the signature for enqueue to:
```
func (c *AvailableConditionController) enqueue(aPIServiceName string) {
```

Reply:
I have checked the code : getAPIServicesFor is called for endpoint related operation as well.